`storage`: rework `compaction_backlog()` logic

WillemKauf opened the pull request:
The previous logic for `disk_log_impl::compaction_backlog()` was quite involved and was mostly predicated on the fact that adjacent `segment` compaction was the only form of compaction in `redpanda` at the time.

To estimate the number of bytes that must be read to fully compact the `log`, search for the last dirty `segment` in the `log`, and sum over all of the bytes that proceed it. This is an over-estimation of the number of bytes that may be processed during compaction, as it does not take into account which keys may be present or not in the dirty/clean `segment`s (ergo which `segment`s will actually be read/rewritten by the built key-offset map during a round of compaction), but this is fine for our purposes.

The `compaction_backlog()` is estimated as 0 if the `log` is not currently eligible for compaction (i.e due to a dirty ratio below `min.cleanable.dirty.ratio` and without a dirty timestamp exceeding `max.compaction.lag.ms`).

This should give a much cleaner and easier to grok estimate of the compaction backlog for a given `log`. The rest of the `compaction_controller` logic is left untouched (i.e backlog is still normalized by the disk capacity, desired backlog size is left as 10%).

## Backports Required

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [X] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none

WillemKauf committed:
`storage`: rework `compaction_backlog()` logic

The previous logic for `disk_log_impl::compaction_backlog()` was quite involved
and was mostly predicated on the fact that adjacent `segment` compaction was
the only form of compaction in `redpanda` at the time.

To estimate the number of bytes that must be read to fully compact the `log`,
search for the last dirty `segment` in the `log`, and sum over all of the bytes
that proceed it. This is an over-estimation of the number of bytes that
may be processed during compaction, as it does not take into account which
keys may be present or not in the dirty/clean `segment`s (ergo which `segment`s
will actually be read/rewritten by the built key-offset map during a round
of compaction), but this is fine for our purposes.

The `compaction_backlog()` is estimated as 0 if the `log` is not currently
eligible for compaction (i.e due to a dirty ratio below `min.cleanable.dirty.ratio`
and without a dirty timestamp exceeding `max.compaction.lag.ms`).

dotnwat commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -3611,145 +3611,31 @@ void disk_log_impl::set_overrides(ntp_config::default_overrides o) {
-    backlog += compaction_backlog_term(std::move(segments_this_term), cf);
-
-    return backlog;
+    return _dirty_segment_bytes;
```
do we read some data twice, for example to create the index then do the compaction scan? how important is it to capture that or is the backlog numbers just normalized against each other and it doens't matter as long as all the ratios reflect the difference in the amount of work?

WillemKauf replied:
> do we read some data twice, for example to create the index then do the compaction scan? how important is it to capture that

I think that the number of dirty bytes on its own is a good reflection of the `compaction_backlog()`. In my opinion, it is _exactly_ representative of the backlog, actually, which is why it makes the most sense as the return value here.

>or is the backlog numbers just normalized against each other

The backlog number is normalized against the disk capacity and there is a set backlog target size of 10%. I think that instead of tweaking the return value to be some linear expression of `_dirty_segment_bytes` (if we wanted to argue that bytes need to be read twice, or that maybe there is some constant associated with that read, so actually the return value should be `2 * _dirty_segment_bytes + C`!, etc...), we would be better off tweaking that target backlog size or the other compaction controller parameters to properly scale the controller behavior.

What do you think?

dotnwat reviewed (dismissed):
why are we backporting? is it fixing a bug....?

WillemKauf commented:
>why are we backporting? is it fixing a bug....?

sort of, its a much better estimate of the compaction backlog than the previous logic which assumed only usage of adjacent merge compaction, which is verging on being a bug due to outdated logic. I think it is also a low risk backport, but let me know if you object heavily to it being backported.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--789e4f7b-b852-4bc8-98b0-2773d137a5b2-->
<!--TEST RESULTS-->
<!--7c3e8582-0cac-4c47-b249-7d4aabcb6e5d-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70092">build#70092</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                                   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:---------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": false, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/70092#019866ce-eaeb-4b81-87d0-980c84393276 | FLAKY         | 20/21    | upstream reliability is '98.92086330935251'. current run reliability is '95.23809523809523'. drift is 3.68277 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70120">build#70120</a></summary>


| test_class                         | test_method                   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------------------------|:------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest          | test_write                    | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70120#01986b25-655d-4090-831b-bf4590a28de3 | FLAKY         | 19/21    | upstream reliability is '93.63143631436314'. current run reliability is '90.47619047619048'. drift is 3.15525 and the allowed drift is set to 50. The test should PASS |
| ControllerLogLimitMirrorMakerTests | test_mirror_maker_with_limits | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70120#01986b27-34fe-48d5-a34a-a53bc56292b9 | FLAKY         | 20/21    | upstream reliability is '99.53917050691244'. current run reliability is '95.23809523809523'. drift is 4.30108 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

andrwng commented:
>the backlog of uncompacted data in the log is equivalent to the number of bytes which have not yet been cleanly compacted.

Hmm, is the backlog supposed to roughly estimate the amount of work compaction is? I kind of buy basing this off dirty data, but I also wonder if we end up in situations where we have a ton of clean data, and only a tiny bit of dirty data -- the amount of IO we have is actually correlated with the total data below the dirty range rather than just the dirty data, since we need to rewrite everything below the dirty data to clean it. Would that end up running through compactions extremely slowly?

>This should give a much cleaner and easier to grok estimate of the compaction backlog for a given log

Is it worth quantifying how far from the ideal number of segments we are as well? Wondering if we could get into a scenario where the log is 100% clean, but needs to be rewritten to merge tiny segments, and we end up running at very low priority?

WillemKauf commented:
> the amount of IO we have is actually correlated with the existing data rather than the new data, since we need to rewrite everything below the dirty data to clean it. Would that end up running through compactions extremely slowly?

This is a very good point. I need to think about this.

WillemKauf commented:
This estimation should be much more involved than I originally posed.

The idea is that we need some idea of the key-set profile in dirty segments, and then on top of that, which segments in the closed portion of the log actually intersect with that dirty key set, leading to something like

```python
backlog = 0
dirty_keys = set()
for segment in segments:
  if not segment.is_clean:
     dirty_keys.update(read_all_keys_from_segment(segment))

for segment in segments:
   segment_keys = read_all_keys_from_segment(segment)
   if segment_keys.intersection(dirty_keys):
     backlog += segment.size()
```

which itself as an estimation function is almost as much read work as the actual compaction. Hmm.

andrwng commented:
> This estimation should be much more involved than I originally posed.
> ...

That does seem pretty involved. Do we gain much by doing the more complicated, more accurate backlog estimation? Conceptually it feels like overestimating up to an extent is fine because if we're wrong, the compaction will finish quicker than expected anyway.

Wondering about something like:

```
last_dirty_seg = None
for segment in reverse(segments):
  if not segment.is_clean:
    last_dirty_seg = segment
    break

backlog = 0
if last_dirty_seg is None:
    return

for segment in segments:
  if segment <= last_dirty_seg:
    backlog += segment.size()
```

But maybe we'd also want to report 0 if the given log has a dirty ratio below `min_cleanable_dirty_ratio` (but also accounting for max lag).

WillemKauf committed:
`storage`: add `log::needs_compaction()`

Factoring out some logic that will be repeated in a future commit
into its own function in the `log`.

WillemKauf commented:
Force push to:
* Add new commit that refactors a section of code into `log::needs_compaction()`.
* Rework logic in `disk_log_impl::compaction_backlog()` to be closer in spirit to the proposed solution(s) above

andrwng commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -4679,4 +4679,16 @@ std::optional<model::offset> disk_log_impl::max_removed_offset() const {
+      = earliest_dirty_ts.has_value()
+        && (now - to_time_point(earliest_dirty_ts.value()) > max_lag);
+
+    auto dr = dirty_ratio();
```
Not suggesting a change or anything, but it's a bummer this isn't const.

WillemKauf replied:
> Not suggesting a change or anything, but it's a bummer this isn't const.

It's a result of `dirty_ratio()` not being `const` (as you probably noticed) to allow us to have a nice safety hatch against negative numbers here as a result of unknown bugs.

https://github.com/redpanda-data/redpanda/blob/24e88437aae0fc14f6dbc143c137abfdd04f6d88/src/v/storage/disk_log_impl.cc#L4535-L4549

We could make `_dirty_segment_bytes` and `_closed_segment_bytes` `mutable`, if you'd prefer that.

WillemKauf replied:
Also, FWIW, it seems like the `dassert()`s in `disk_log_impl::subtract_dirty_segment_bytes()` [1] and `disk_log_impl::subtract_closed_segment_bytes()` haven't caused any CI failures, so I'm fairly confident this safety hatch is unnecessary at this point, but as a defensive coding measure I still appreciate having it around.

[1]: 

https://github.com/redpanda-data/redpanda/blob/24e88437aae0fc14f6dbc143c137abfdd04f6d88/src/v/storage/disk_log_impl.cc#L4567-L4580

andrwng replied:
Yea I can get behind making it `mutable`, though don't feel particularly strongly. Just thought i'd call this out because it made me do a double take

WillemKauf replied:
https://github.com/redpanda-data/redpanda/pull/27114

vbotbuildovich commented:
/backport v25.2.x

vbotbuildovich commented:
Failed to create a backport PR to v25.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-27099-v25.2.x-217 remotes/upstream/v25.2.x
git cherry-pick -x 7ecead4287 d0a6187de6
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/16730019353).
